### PR TITLE
Adds `yii\base\Module::getReflection()` and updates related methods

### PR DESCRIPTION
**Pull request by WinterSilence**
Adds method `yii\base\Module::getReflection()` and updates related methods.

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  |

### CODE REVIEW

**Comment by samdark** [comment]
What is the purpose of this change?

File: framework/base/Module.php [hunk 1]
```diff
@@ -143,6 +144,11 @@ class Module extends ServiceLocator
      * @since 2.0.11
      */
     private $_version;
+    /**
+     * @var ReflectionObject|null the reflection instance for this class
+     * @since 2.0.44
```
[the remark this answers is not in the document]
Comment (WinterSilence) [reply framework/base/Module.php:149]: @samdark  now I'm understand you, removed

File: framework/base/Module.php [hunk 2]
```diff
@@ -195,14 +201,27 @@ public static function setInstance($instance)
...
+            $this->controllerNamespace = $this->getReflection()->getNamespaceName() . '\controllers';
         }
     }
 
+    /**
+     * Returns reflection instance for this class.
+     * @return ReflectionObject
+     * @since 2.0.44
+     */
+    protected function getReflection()
```
[the remark this answers is not in the document]
Comment (WinterSilence) [reply framework/base/Module.php:215]: @samdark now code easy and pure. API - is public methods. 

> We aren't using it and I don't think that end users will call

It's used in that's class, don't decide for others

[the remark this answers is not in the document]
Comment (samdark) [reply framework/base/Module.php:215]: But that's pure internals and I don't see any usage for it except in `getBasePath()`.

[the remark this answers is not in the document]
Comment (WinterSilence) [reply framework/base/Module.php:215]: @samdark this does not mean that users don't see too

**Comment by samdark** [comment]
Won't merge till more votes for adding it. I don't see `getReflection()` useful for cases other than current one.

**Comment by WinterSilence** [comment]
@samdark what's about moving `getReflection()` into `Component`? then we can update
https://github.com/yiisoft/yii2/blob/4c1f872320310ad719e660f183c5dc91088463b1/framework/di/Container.php#L508
https://github.com/yiisoft/yii2/blob/4c1f872320310ad719e660f183c5dc91088463b1/framework/base/Controller.php#L257
https://github.com/yiisoft/yii2/blob/4c1f872320310ad719e660f183c5dc91088463b1/framework/base/Model.php#L255
https://github.com/yiisoft/yii2/blob/4c1f872320310ad719e660f183c5dc91088463b1/framework/base/Model.php#L270
and etc.

**Comment by rob006** [comment]
You can't serialize `ReflectionClass`, so this is no-go for `Model`, and I would argue that it is not worth the trouble for `Module` either.

**Comment by WinterSilence** [comment]
@rob006
> You can't serialize ReflectionClass

Why? it's work fine:
~~~php
$class = new \ReflectionClass(\yii\db\ActiveRecord::class);
$serializedClass = serialize($class);
~~~
or clear property at serialization:
~~~php
public function __sleep()
{
        $this->_reflection = null;
        return array_keys(get_object_vars($this));
}
~~~

**Comment by rob006** [comment]
> Why? it's work fine:

It does not work for PHP 7.4+

https://3v4l.org/gbDbZ

**Comment by WinterSilence** [comment]
@rob006 then **plan b**: clean reflection property at serialization

**Comment by rob006** [comment]
That would be a BC break if someone is already implementing `__sleep()`. And benefits of having such getter is negligible (and it is actually a drawback if we make it protected).

**Comment by WinterSilence** [comment]
@rob006
> That would be a BC break if someone is already implementing __sleep().

then impossible add **any** method/class - someone has definitely implemented it and `getReflection()` too (: 

Ok, we can store reflection instance only if php < 7.4, for modern PHP we create reflection instance at every call.
Method is protected because in current classes that's using for self reflection. If move method to `Component` then any component you can call method of another component

**Comment by samdark** [comment]
btw., do we have any numbers confirming positive impact on performance/memory usage?

**Comment by WinterSilence** [comment]
@samdark this method was added not for optimization purposes, but to simplify access to self reflection. It's refactoring: I extracted the duplicate code into method.

**Comment by samdark** [comment]
I know but previously reflection instance wasn't stored. Now it is and it should somehow affect performance / memory usage.

**Comment by WinterSilence** [comment]
@samdark plus: speed up - no need create object every time,  minus: memory usage

**Comment by rob006** [comment]
I don't think there is any measurable performance difference - reflection is used 2 times at most, so caching reflection object does not change that much.

**Comment by WinterSilence** [comment]
@rob006 
> this method was added not for optimization purposes, but to simplify access to self reflection. It's refactoring: I extracted the duplicate code into method.

**Comment by rob006** [comment]
You extracted `new ReflectionObject($this)` - calling such simple instantiation as "duplicated code" is ridiculous. Not to mention that it was you who created such "duplication" - original code in `init()` did not used reflection.

**Comment by WinterSilence** [comment]
@rob006 yes, but worked slower

**Comment by samdark** [comment]
@WinterSilence how slower? What percentage?

**Comment by WinterSilence** [comment]
@samdark x2
~~~php
use yii\db\ActiveRecord;
$i = 0;
$time = -microtime(true);
while ($i < 500) {
    $class = get_class(new ActiveRecord());
    if (($pos = strrpos($class, '\\')) !== false) {
        $value = substr($class, 0, $pos) . '\\controllers';
    }
    $i++;
}
echo sprintf('%f', $time + microtime(true)); // 0,022300

$i = 0;
$time = -microtime(true);
while ($i < 500) {
    $value = (new ReflectionClass(new ActiveRecord()))->getNamespaceName() . '\controllers';
    $i++;
}
< sprintf('%f', $time + microtime(true)); // 0,009237
~~~

**Comment by rob006** [comment]
There is no noticeable difference: https://3v4l.org/6ZH18

**Comment by samdark** [comment]
@WinterSilence that's because you test in the same file so when it reaches a second case `ActiveRecord` is already resolved and cached. In order to have fair comparison these should be at least in separate files.

**Comment by samdark** [comment]
@rob006 same thing happens in your case.

**Comment by rob006** [comment]
@samdark There is no autoloading in my case. What do you mean by "resolved"?

**Comment by samdark** [comment]
Split your tests into separate files and try.

**Comment by WinterSilence** [comment]
@samdark I'm understand you at first comment, but I don't know how to explain it purest:
> this method was added not for optimization purposes, but to simplify access to self reflection.

**Comment by samdark** [comment]
Despite the fact, it may affect performance so it should be measured and measured at least w/o side effects. Ideally, a phpbench since it is getting average and counts delta, but separate files should be alright.

**Pull request closed without merging**